src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,256 @@ namespace wasm {
+      moveableExpressions.erase(moveableExpressions.begin(),
+                                moveableExpressions.begin() + minIndex);
+      earliestRefAsEligible -= minIndex;
+      earliestRefCastEligible -= minIndex;
+    }
+  }
+
+  void visitLocalGet(LocalGet* curr, Module& module) {
+    moveableExpressions.push_back(curr);
+    EffectAnalyzer currEffectAnalyzer(options, module);
+    currEffectAnalyzer.visit(curr);
+    effectAnalyzers.emplace(curr, currEffectAnalyzer);
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    moveableExpressions.clear();
+    effectAnalyzers.clear();
+    earliestRefAsEligible = 0;
+    earliestRefCastEligible = 0;
+  }
+
+  void visitRefAs(RefAs* curr, LocalGet* fallthrough) {
+    auto iter = refAsReplacement.find(moveableExpressions.front());
+    if (iter == refAsReplacement.end()) {
+      // Since RefAs ops are orthogonal to each other, we currently just
+      // select the first RefAs we come accross which is moveable.
+      refAsReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefAsReplacement.insert(fallthrough);
+      }
+    }
+  }
+
+  void visitRefCast(RefCast* curr, LocalGet* fallthrough) {
+    auto iter = refCastReplacement.find(moveableExpressions.front());
+    if (iter == refCastReplacement.end()) {
+      refCastReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefCastReplacement.insert(fallthrough);
+      }
+    } else if (curr->type != iter->second->type &&
+               Type::isSubType(curr->type, iter->second->type)) {
+      iter->second = curr;
+      if (fallthrough != iter->first) {
+        redundantRefCastReplacement.erase(iter->first);
+      }
+    }
+  }
+
+  void removeRedundantReplacements() {
+    for (auto& redundantRefAs : redundantRefAsReplacement) {
+      refAsReplacement.erase(redundantRefAs);
+    }
+    for (auto& redundantRefCast : redundantRefCastReplacement) {
+      refCastReplacement.erase(redundantRefCast);
+    }
+  }
+};
+
+// Find the "best" cast to move earlier to another local.gets
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+
+  // For each local index, we have an searcher to move casts for it.
+  std::unordered_map<Index, IndexEarlyCastSearcher> castMoveSearchers;
+
+  // Maps local.gets to the "best" RefAs to move to it.
+  std::unordered_map<Expression*, RefAs*> refAsToApply;
+
+  // Maps local.gets to the "best" RefCast to move to it.
+  std::unordered_map<Expression*, RefCast*> refCastToApply;
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    self->castMoveSearchers.clear();
+  }
+
+  void visitExpression(Expression* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
```

Comment:
```suggestion
    for (auto& [_, searcher] : castMoveSearchers) {
      searcher.visitExpression(curr);
```
I think this style is clearer, and we use it a lot. Note how it gets rid of the `.second` which can be a little unclear when reading code.

Reply:
Not in new change.